### PR TITLE
Reject queries that are recursive during the resolution

### PR DESCRIPTION
**Pull request by xgreenx**
If during resolution of the query we hit the same type, reject further execution. The code to do this check uses `Visitor` logic from the `async-graphql` library, but this logic is private, so I duplicated it in our branch, Later it would be nice to create PR to them in make it public. Or add a way to pass our own query validators into the  `async-graphql`(which will be more performant).

## Checklist
- [x] Breaking changes are clearly marked as such in the PR description and changelog
- [x] New behavior is reflected in tests

### Before requesting review
- [x] I have reviewed the code myself

### CODE REVIEW

File: Cargo.toml
Comment (netrome) [inline Cargo.toml:151]: Nit: I assume we don't want to commit this part.
```suggestion
```

File: crates/fuel-core/src/graphql_api/validation_extension.rs
Comment (netrome) [inline crates/fuel-core/src/graphql_api/validation_extension.rs:41]: Is this so performance critical that we need the parking lot mutex instead of the std one? If we introduce parking lot, I assume we'd likely want to follow up with using their synchronization primitives everywhere (which could be a separate issue).

File: crates/fuel-core/src/graphql_api/validation_extension.rs
Comment (netrome) [inline crates/fuel-core/src/graphql_api/validation_extension.rs:56]: Houston, we have a problem. This extension is shared between all concurrent query tasks. So if we have multiple requests for the same field, we may accidentally reject one query as recursive.

I tested it with this scenario:

```rust
#[tokio::test]
async fn submitting_multiple_queries_should_not_fail_but_it_does() {
    let query = r#"
    query {
      blocks(last: 1) {
        edges {
          cursor
          node {id}
        }
      }
    }"#;

    let node = FuelService::new_node(Config::local_node()).await.unwrap();
    let url = format!("http://{}/v1/graphql", node.bound_address);
    let client = FuelClient::new(url.clone()).unwrap();

    client
        .submit_and_await_commit(&Transaction::default_test_tx())
        .await
        .unwrap();

    for _ in 0..2 {
        let url = url.clone();
        let handle = tokio::spawn(async move {
            let result = send_graph_ql_query(&url, query).await;
            assert!(!result.contains("Recursion detected"), "{:?}", result);
        });
    }
}
```

which fails when inserting a ```tokio::time::sleep(std::time::Duration::from_millis(100)).await;``` before the `next.run` function.

Unfortunately I don't see a simple fix for this. We could perhaps do something with `tokio::task::id` ([link](https://docs.rs/tokio/latest/tokio/task/fn.id.html)) but that is an unstable feature.

Alternatively, this seems pretty unlikely (I could not trigger the error without inserting the artificial `sleep` call), so we could accept the risk for now and create a follow-up issue for the proper fix.

**Review by netrome — COMMENTED** [review]
Spotted a race condition that can cause false positive recursion detections. It's not obvious to me whether this is a showstopper or not depending on how pragmatic we want to be. Let me know your thoughts. Meanwhile, I'll think about potential ways to handle this problem.

File: crates/fuel-core/src/graphql_api/validation_extension.rs
Comment (xgreenx) [reply crates/fuel-core/src/graphql_api/validation_extension.rs:56]: It fails indeed, but not for the reason that you've mentioned=D

Each query has its own copy of the extension, so parallel query will have its own `visited: parking_lot::Mutex<HashMap<String, usize>>`.

But within one query, resolvers for other fields start to run when the resolver future is paused. And it is why test fails=)

```rust
#[tokio::test]
async fn submitting_multiple_queries_should_not_fail_but_it_does() {
    let query = r#"
    query {
      blocks(last: 1) {
        edges {
          cursor
          node {id}
        }
      }
    }"#;

    let node = FuelService::new_node(Config::local_node()).await.unwrap();
    let url = format!("http://{}/v1/graphql", node.bound_address);
    let client = FuelClient::new(url.clone()).unwrap();

    client
        .submit_and_await_commit(&Transaction::default_test_tx())
        .await
        .unwrap();
}
```

Comment (netrome) [reply crates/fuel-core/src/graphql_api/validation_extension.rs:56]: Ah yeah you're right. There's no Arc, so there is no way the Mutex is shared between tasks. But still it's concurrently accessed like you say.

Comment (xgreenx) [reply crates/fuel-core/src/graphql_api/validation_extension.rs]: Fixed with one time analyze during parsing of the query

File: crates/fuel-core/src/graphql_api/validation_extension/visitor.rs
Comment (netrome) [inline crates/fuel-core/src/graphql_api/validation_extension/visitor.rs:1]:
```suggestion
//! The code below is copy from `<https://github.com/async-graphql/async-graphql/blob/da5fea4023b8b043c5295f0a13ba76180d9e2c53/src/validation/visitor.rs>`.
```

File: crates/fuel-core/src/graphql_api/validation_extension.rs
Comment (netrome) [inline crates/fuel-core/src/graphql_api/validation_extension.rs:50]: Do we need the option here? I.e. is it meaningful to differentiate between an empty vec of errors and no errors? Otherwise I think we can simplify this to
```suggestion
    errors: Mutex<Vec<RuleError>>,
```

File: crates/fuel-core/src/graphql_api/validation_extension/visitor.rs
Comment (netrome) [reply crates/fuel-core/src/graphql_api/validation_extension/visitor.rs:1]: Also, ugh. Would be really nice to expose this upstream!

**Review by netrome — APPROVED** [review]
Nice stuff, thanks for fixing!

**Review by rymnc — COMMENTED** [review]
can we use the image generated by this pr on devnet before merging?

**Review by rymnc — DISMISSED** [review]
maybe we need a review from @Dentosal to be able to merge.

approving so the new release isn't blocked

**Review by Dentosal — APPROVED** [review]
The code looks fine. I'm not fan of this idea at all, as having GraphQL without recursive ("graph") queries is nonsensical and just worse than a normal REST API. Have we really carried all this complexity for nothing? But given the timelines, this is the practical solution so I'm approving it anyway.

**Review by rymnc — APPROVED** [review]

**Review by Dentosal — APPROVED** [review]

**Pull request merged**
